lib/files/index.js

```diff
@@ -0,0 +1,119 @@
+
+const filesInterface = (app, settings) => {
+    const BASE_PATH = join(settings.rootDir, settings.userDir, 'storage')
+    // not sure about this yet, need to check how file storage works
+    const storage = multer.memoryStorage()
+    const upload = multer({ storage })
+
+    const regex = /^\/flowforge\/files\/_\/(.*)$/
+
+    const listDirectory = async (path = '') => {
+        const fullPath = normalize(join(BASE_PATH, path))
+        if (fullPath.startsWith(BASE_PATH)) {
+            const files = await readdir(fullPath, { withFileTypes: true })
+            const response = {
+                meta: { },
+                files: [],
+                count: files.length
+            }
+            for (const i in files) {
+                const file = files[i]
+                const fullPath = join(file.path, file.name)
+                const s = await stat(fullPath)
+                const rep = {
+                    name: file.name,
+                    lastModified: s.mtime
+                }
+                if (file.isFile()) {
+                    rep.type = 'file'
+                    rep.size = s.size
+                } else if (file.isDirectory()) {
+                    rep.type = 'directory'
+                }
+                response.files.push(rep)
+            }
+            return response
+        } else {
+            return []
+        }
+    }
+
+    app.get(regex, async (request, reply) => {
+        try {
+            const files = await listDirectory(request.params[0])
+            reply.send(files)
+        } catch (err) {
+            if (err.code === 'ENOENT') {
+                reply.status(404).send()
+            } else if (err.code === 'ENOTDIR') {
+                reply.status(400).send()
+            }
+        }
+    })
+
+    app.put(regex, async (request, reply) => {
+        const fullPath = normalize(join(BASE_PATH, request.params[0]))
+        const newPath = normalize(join(BASE_PATH, request.body.path))
+        if (fullPath.startsWith(BASE_PATH) && newPath.startsWith(BASE_PATH)) {
+            if (existsSync(fullPath) && existsSync(dirname(newPath))) {
+                await rename(fullPath, newPath)
+                reply.status(202).send()
+            } else {
+                reply.status(404).send()
+            }
+        }
+        reply.status(403).send()
+    })
+
+    app.post(regex, upload.single('file'), async (request, reply) => {
+        const startPath = normalize(join(BASE_PATH, request.params[0]))
+        console.log(startPath)
+        if (startPath.startsWith(BASE_PATH)) {
+            console.log(request.get('content-type'))
+            if (request.get('content-type') === 'application/json') {
+                const newPath = request.body.path
+                const fullPath = normalize(join(startPath, newPath))
+                console.log(fullPath)
+                if (fullPath.startsWith(BASE_PATH)) {
+                    await mkdir(fullPath, { recursive: true })
```

Comment:
If I have `foo/file.txt` on disk already, and a request comes to create a directory `foo/file.txt/blah`, this throws an error which goes uncaught and crashes the launcher (because file.txt is not a directory that can have subdirs).

All of the `await` calls in *all* route handlers must have a try/catch to guard against unexpected errors.